fedimint-core/src/module/mod.rs

```diff
@@ -851,6 +851,13 @@ pub trait ServerModule: Debug + Sized {
         peer_id: PeerId,
     ) -> anyhow::Result<()>;
 
+    fn verify_input(
+        &self,
+        _input: &<Self::Common as ModuleCommon>::Input,
+    ) -> Result<(), <Self::Common as ModuleCommon>::InputError> {
+        Ok(())
+    }
```

Comment:
This needs *a lot* more explanation. The semantics as I understand them:

* Anything that can be checked about an input as a pure function (i.e. *without* reading from the DB/any global state) and is computationally heavy enough to matter should be checked using `verify_input`.
  * It is guaranteed that `verify_input` is run before `process_input` and that `process_input` is not run if it fails.
  * For most modules `verify_input` can be ignored since their stateless checks are supposedly fast compared to their stateful check.
* Anything that interacts with the DB (e.g. spent-ness checks for e-cash, account balance checks for LN, …) still has to be checked by `process_input` since that runs serially inside the database transaction and can ensure there are no race conditions etc.

**Doing stateful checks inside `verify_input` will nearly always lead to inflation bugs, so should be avoided at all cost!**